refactor(compiler-cli): include filepath for `DocEntry`

JeanMeche opened the pull request:
This is to allow the API docs to provide a direct link to implementation on github.

Here is a screen of what info we get. 
![Screenshot 2024-04-03 at 22 04 23](https://github.com/angular/angular/assets/1300985/92a29376-580b-491e-be44-d0ff5d44e521)


vs what aio provides: https://github.com/angular/angular/tree/17.3.2/packages/common/src/pipes/date_pipe.ts#L60-L281 

Related to #52641

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:
```diff
@@ -46,8 +46,14 @@ export class DocsExtractor {
+        endLine: ts.getLineAndCharacterOfPosition(node.getSourceFile(), node.getEnd()).line,
+      };
+
+      const entry = this.extractDeclaration(node, source);
```
Instead of passing it down to the extractors, we can convert a `DocEntry` to a `DocEntryWithSource` potentially.

JeanMeche replied:
I don't think introducing `DocEntryWithSource ` will be ideal because of the inheritance. I have refactored to prevent passing down the source though. Let me know what you think of it.

devversion replied:
What cases of inheritance do you think of? Using `Omit` all the time seems non-ideal.

I would imagine something like

```ts
interface DocEntryWithSourceInfo extends DocEntry {
  ..
}

function isDocEntryWithSource(d: DocEntry): d is DocEntryWithSourceInfo {
  ...
}
```

and then when needed, the source doc info can be used if available.

JeanMeche replied:
Oh I understand your idea now. My worry was that every entity was extending `DocEntry`. 
Creating a completly separate interface makes it indeed simplier. Thank you.

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:
```diff
@@ -46,8 +46,14 @@ export class DocsExtractor {
+      const source: SourceEntry = {
+        filePath: getRelativeFilePath(node.getSourceFile()),
+        startLine: ts.getLineAndCharacterOfPosition(node.getSourceFile(), node.getFullStart()).line,
+        endLine: ts.getLineAndCharacterOfPosition(node.getSourceFile(), node.getEnd()).line,
```
Nit: use `sourceFile` in all places

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:
```diff
@@ -157,3 +174,13 @@ function isIgnoredDocEntry(entry: DocEntry): boolean {
+  const fullPath = sourceFile.fileName;
+
+  // we want to remove the part before /packages
+  return fullPath.replace(/.*?\/packages/, '/packages');
```
Can we not hard-code this and use the program rootDir? The extraction is not specific to framework, necessarily

JeanMeche committed:
refactor(compiler-cli): include filepath for `DocEntry`

This is allow the API docs to provide a direct link to implementation on github.

devversion commented on packages/compiler-cli/src/ngtsc/docs/src/extractor.ts:
```diff
@@ -157,3 +164,18 @@ function isIgnoredDocEntry(entry: DocEntry): boolean {
+  const fullPath = sourceFile.fileName;
+
+  // What matters is in the directory located inside the execroot
+  const relativePath = fullPath.match(/execroot\/[^/]+\/(.+)/)?.[1];
```
Wouldn't it be possible to use the program `rootDir` and not assume execroot etc?

JeanMeche replied:
I wasn't sur what `rootDir` you were referencing (`import.meta.url` wasn't useful for example) but I think I got it !

devversion commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -734,7 +734,8 @@ export class NgCompiler {
     }
 
-    return docsExtractor.extractAll(entryPointSourceFile);
+    const rootDir = this.inputProgram.getCurrentDirectory();
```
Technically the current directory is not the root dir. We can leave this as is for now, but ideally it's derived differently from the config. I think ngtsc does this somewhere for the flat entry point resolution

JeanMeche replied:
I've added a TODO. 

For the context we almost have a circular dependency between this PR and angular/dev-infra#1957. 
I'll be happy to do a follow-up once both this one and the dev-infra one are merged and we can have easy testing of the output of this function.

devversion reviewed (approved):
LGTM- but we are not picking up the 100% correct root dir

pkozlowski-opensource commented:
This PR was merged into the repository by commit 9afa2ea345df6fdb0db6e69f1e8a69f2c489f4e4.
